### ISSUE TITLE
Collaboration model using project nodes

### ISSUE THREAD

**Issue by nicknikolov** [pex-gl/pex-nodes#198]
So I have an idea.

## The model
First let's name custom nodes that you downloaded to disk **project nodes**.
Now, every time you download it (let's call this perhaps *save to project*?, I don't like how downlaod sounds) you create a **version**. This will be perhaps just the code, without data, timestamped and will have some unique id.
Now here is a scenario:
1. I create a graph. There I create a custom node called "Nick". I work on it locally (meaning I keep it in graph.json). When I am happy with it I "save it to project". I created the first version of "Nick".

2. You clone my graph or create another one. You like my node so you create an instance of it from the node picker. You find a bug or extend the functionality of "Nick" by changing the source code and keeping it inside graph.json. At some point you decide that it works the way you like it and save it to project. Now what happens is the second version of "Nick" is created. Nothing else gets changed in the project, except that a new version exists and you use it in place.

3. I open my original graph and pex-nodes lets me know that I have a project node that has a new version. Upgrading can be as simple as having a dropdown of versions in the inspector and I can click the latest one to change the source code, this will allow me to quickly go back to an older version if something breaks. And done! We collaborated without any breaking changes, you prototyped and created a new node locally in your graph, I upgraded when I had the chance to test and approve your version.

## The big picture.

When you think about it, using this workflow we could create a master graph containing the whole app (which we usually have anyway). But this master graph could in theory act as a "test suite", and most parts of it can be cloned in seperate graphs where we work on the functionality of that chunk of the graph and nothing else. When we have new features we push them via new version of project nodes. Then someone just opens the master graph and "patches" those changes, testing that everything works as before. This could be very productive for 1 dev and also for 10. You can also think of it as similar to most git workflows, where you develop every feature in a branch and merge back to master when ready. 

Thoughts?

**Comment by vorg** [comment]
I agree we need versioning. This would also allow us to version core nodes (i think cables does it too). What I'm not sure is manual patching. The changes are usually more than code updated for few nodes but also param values, connections etc.  In the end we need groups / subgraphs I think. Can I save group node as custom node? Effectively creating an importable subgraph? What defines node version continuity? Is it just the same name? Sketching some cases / user scenarios could be a next step.

**Comment by nicknikolov** [comment]
> The changes are usually more than code updated for few nodes but also param values, connections etc

This is why I said the nodes keep it's own data. E.g. when I set new version of my node, I don't use data. If the new version has changed ports and connection get lost then you either fix (there is no other solution really, if you change API of the node then it will be a breaking change and its up to you to fix, just like any other software upgrade) or you keep old version. I think this approach is good because you control the process, no magic behind the scenes. 

> In the end we need groups / subgraphs I think
Sure but I don't think this is related to my proposal. Either way, I will want to collaborate on a node level, not only with groups/subgraphs.

> What defines node version continuity? Is it just the same name?

I was thinking for a better identifier but I think yes.. just like npm uses name as a unique id. 

Try reading my proposal again for user case and scenarios but I'll try to think up some more later.

**Comment by nicknikolov** [comment]
Ok I understand what you meant by subgraphs now. E.g you change the graph structure itself  - new nodes, delete some nodes, rewire things etc. Yeah sure, collaborating on a subgraph level like that might work but again - this is a different problem which will probably be solved in a different manner. I'm focusing solely on node-level collab here.
